sql/rowexec: TestDrainingProcessorSwallowsUncertaintyError failed

[(sql/rowexec).TestDrainingProcessorSwallowsUncertaintyError failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2015943&tab=buildLog) on [master@82538176c897591e8241fa258895e42b9a3226a4](https://github.com/cockroachdb/cockroach/commits/82538176c897591e8241fa258895e42b9a3226a4):

```
I200616 18:22:21.808098 17129 server/server.go:1442  [n3] starting grpc server at 127.0.0.1:44291
I200616 18:22:21.808119 17129 server/server.go:1443  [n3] advertising CockroachDB node at 127.0.0.1:44291
I200616 18:22:21.813336 18617 server/server.go:1070  active cluster version is now 20.1-7 (up from 20.1)
I200616 18:22:21.833336 18650 sql/temporary_schema.go:466  [n3] skipping temporary object cleanup run as it is not the leaseholder
I200616 18:22:21.833448 18650 sql/temporary_schema.go:467  [n3] completed temporary object cleanup job
I200616 18:22:21.833467 18650 sql/temporary_schema.go:595  [n3] temporary object cleaner next scheduled to run at 2020-06-16 18:52:21.831935927 +0000 UTC
I200616 18:22:21.855757 17129 server/server_sql.go:645  [n3] done ensuring all necessary migrations have run
I200616 18:22:21.855977 17129 server/server.go:1756  [n3] serving sql connections
I200616 18:22:21.857818 18692 sqlmigrations/migrations.go:727  [n3] starting wait for upgrade finalization before schema change job migration
I200616 18:22:21.867891 18692 sqlmigrations/migrations.go:753  [n3] schema change job migration already complete
I200616 18:22:21.906504 18648 server/server_update.go:55  [n3] no need to upgrade, cluster already at the newest version
I200616 18:22:21.946907 18649 sql/event_log.go:132  [n3] Event: "node_join", target: 3, info: {Descriptor:{NodeID:3 Address:127.0.0.1:44291 Attrs: Locality:region=test,dc=dc3 ServerVersion:20.1-7 BuildTag:v20.2.0-alpha.00000000-2370-g82538176c8 StartedAt:1592331741803636822 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:42053} ClusterID:8635a6aa-b683-47da-a179-d19c43cdea4f StartedAt:1592331741803636822 LastUp:1592331741803636822}
I200616 18:22:21.950649 18287 sql/event_log.go:132  [n1,client=127.0.0.1:58578,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I200616 18:22:21.991782 18287 sql/event_log.go:132  [n1,client=127.0.0.1:58578,hostssl,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I200616 18:22:22.010134 18287 sql/event_log.go:132  [n1,client=127.0.0.1:58578,hostssl,user=root] Event: "create_table", target: 53, info: {TableName:test.public.t Statement:CREATE TABLE test.public.t (x INT8 PRIMARY KEY) User:root}
I200616 18:22:22.028928 18287 kv/kvserver/replica_command.go:413  [n1,s1,r34/1:/{Table/38-Max}] initiating a split of this range at key /Table/53/1/6 [r35] (manual)
I200616 18:22:22.041766 18287 kv/kvserver/replica_command.go:1605  [n1,s1,r35/1:/{Table/53/1/6-Max}] change replicas (add [(n2,s2):2LEARNER] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, next=2, gen=1, sticky=9223372036.854775807,2147483647]
I200616 18:22:22.050445 18287 kv/kvserver/replica_raft.go:247  [n1,s1,r35/1:/{Table/53/1/6-Max}] proposing SIMPLE(l2) ADD_REPLICA[(n2,s2):2LEARNER]: after=[(n1,s1):1 (n2,s2):2LEARNER] next=3
I200616 18:22:22.052333 18287 kv/kvserver/store_snapshot.go:988  [n1,s1,r35/1:/{Table/53/1/6-Max}] streamed LEARNER snapshot 04ccfa33 at applied index 14 to (n2,s2):2LEARNER in 0.00s @ 0 B/s: kv pairs: 11, log entries: 0, rate-limit: 8.0 MiB/s, queued: 0.00s
I200616 18:22:22.052765 18944 kv/kvserver/raft_snapshot_queue.go:128  [n1,raftsnapshot,s1,r35/1:/{Table/53/1/6-Max}] skipping snapshot; replica is likely a learner in the process of being added: (n2,s2):2LEARNER
I200616 18:22:22.052870 19059 kv/kvserver/replica_raftstorage.go:791  [n2,s2,r35/2:{-}] applying LEARNER snapshot [id=04ccfa33 index=14]
I200616 18:22:22.054139 19059 kv/kvserver/replica_raftstorage.go:812  [n2,s2,r35/2:/{Table/53/1/6-Max}] applied LEARNER snapshot [total=1ms ingestion=4@1ms id=04ccfa33 index=14]
I200616 18:22:22.055107 18287 kv/kvserver/replica_command.go:1605  [n1,s1,r35/1:/{Table/53/1/6-Max}] change replicas (add [(n2,s2):2] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, (n2,s2):2LEARNER, next=3, gen=2, sticky=9223372036.854775807,2147483647]
I200616 18:22:22.058267 18287 kv/kvserver/replica_raft.go:247  [n1,s1,r35/1:/{Table/53/1/6-Max}] proposing SIMPLE(v2) ADD_REPLICA[(n2,s2):2]: after=[(n1,s1):1 (n2,s2):2] next=3
I200616 18:22:22.111980 19083 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove [(n1,s1):1VOTER_DEMOTING]): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]
I200616 18:22:22.122075 19083 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing ENTER_JOINT(r1 l1) REMOVE_REPLICA[(n1,s1):1VOTER_DEMOTING]: after=[(n1,s1):1VOTER_DEMOTING (n2,s2):2] next=3
I200616 18:22:22.123328 19083 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1VOTER_DEMOTING, (n2,s2):2, next=3, gen=4, sticky=9223372036.854775807,2147483647]
I200616 18:22:22.132091 19083 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing LEAVE_JOINT: after=[(n1,s1):1LEARNER (n2,s2):2] next=3
I200616 18:22:22.133394 19083 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove [(n1,s1):1LEARNER]): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1LEARNER, (n2,s2):2, next=3, gen=5, sticky=9223372036.854775807,2147483647]
I200616 18:22:22.142107 19083 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing SIMPLE(r1) REMOVE_REPLICA[(n1,s1):1LEARNER]: after=[(n2,s2):2] next=3
I200616 18:22:22.143022 17216 kv/kvserver/store_remove_replica.go:122  [n1,s1,r35/1:/{Table/53/1/6-Max}] removing replica r35/1
I200616 18:22:22.151980 18287 sql/event_log.go:132  [n1,client=127.0.0.1:58578,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.defaults.results_buffer.size Value:0 User:root}
--- FAIL: TestDrainingProcessorSwallowsUncertaintyError (1.12s)
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=false
        --- FAIL: TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=false (0.00s)
            processors_test.go:671: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1592331742.319910530,0 encountered previous write with future timestamp 1592331742.319910531,0 within uncertainty interval `t <= 1592331742.819910530,0`; observed timestamps: [{1 1592331742.319910530,0}] (SQLSTATE 40001)
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=true
I200616 18:22:22.329610 17129 util/stop/stopper.go:539  quiescing
I200616 18:22:22.329667 19320 util/stop/stopper.go:539  quiescing
I200616 18:22:22.329804 19318 util/stop/stopper.go:539  quiescing
I200616 18:22:22.329903 19319 util/stop/stopper.go:539  quiescing
W200616 18:22:22.338833 18764 kv/kvserver/intentresolver/intent_resolver.go:744  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
W200616 18:22:22.341736 19076 kv/kvserver/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 1: EOF:
W200616 18:22:22.342366 18933 kv/kvserver/raft_transport.go:637  [n1] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200616 18:22:22.347384 19202 kv/kvserver/intentresolver/intent_resolver.go:744  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/53/1/6/RangeDescriptor: node unavailable; try another peer
W200616 18:22:22.348208 18938 kv/kvserver/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
        --- FAIL: TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=true (0.01s)
            processors_test.go:675: expected injected error, got: ERROR: rpc error: code = Canceled desc = context canceled (SQLSTATE XXUUU)
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=true
    --- FAIL: TestDrainingProcessorSwallowsUncertaintyError/vectorize=true (0.01s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDrainingProcessorSwallowsUncertaintyError PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDrainingProcessorSwallowsUncertaintyError.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/rowexec).TestDrainingProcessorSwallowsUncertaintyError failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2016276&tab=buildLog) on [master@82538176c897591e8241fa258895e42b9a3226a4](https://github.com/cockroachdb/cockroach/commits/82538176c897591e8241fa258895e42b9a3226a4):

```
I200616 19:18:47.967340 18204 server/server.go:1339  [n?] node connected via gossip
I200616 19:18:47.967432 18203 gossip/server.go:227  [n1] received initial cluster-verification connection from 127.0.0.1:45571
I200616 19:18:48.031711 16901 server/node.go:365  [n?] new node allocated ID 3
I200616 19:18:48.031772 16901 gossip/gossip.go:406  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:45571" > attrs:<> locality:<tiers:<key:"region" value:"test" > tiers:<key:"dc" value:"dc3" > > ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:7 > build_tag:"v20.2.0-alpha.00000000-2370-g82538176c8" started_at:1592335128031761398 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:46365" > 
I200616 19:18:48.032142 16901 kv/kvserver/stores.go:236  [n3] read 0 node addresses from persistent storage
I200616 19:18:48.032347 16901 kv/kvserver/stores.go:255  [n3] wrote 2 node addresses to persistent storage
I200616 19:18:48.033091 16904 kv/kvserver/stores.go:255  [n1] wrote 2 node addresses to persistent storage
I200616 19:18:48.033449 17842 kv/kvserver/stores.go:255  [n2] wrote 2 node addresses to persistent storage
I200616 19:18:48.034950 16901 server/node.go:568  [n3] bootstrapped store [n3,s3]
I200616 19:18:48.035216 16901 server/node.go:468  [n3] node=3: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200616 19:18:48.035862 16901 server/server.go:1435  [n3] starting https server at 127.0.0.1:38825 (use: 127.0.0.1:38825)
I200616 19:18:48.035916 16901 server/server.go:1440  [n3] starting postgres server at 127.0.0.1:46365 (use: 127.0.0.1:46365)
I200616 19:18:48.035943 16901 server/server.go:1442  [n3] starting grpc server at 127.0.0.1:45571
I200616 19:18:48.035963 16901 server/server.go:1443  [n3] advertising CockroachDB node at 127.0.0.1:45571
I200616 19:18:48.053164 18420 server/server.go:1070  active cluster version is now 20.1-7 (up from 20.1)
I200616 19:18:48.077014 18173 sql/temporary_schema.go:466  [n3] skipping temporary object cleanup run as it is not the leaseholder
I200616 19:18:48.077130 18173 sql/temporary_schema.go:467  [n3] completed temporary object cleanup job
I200616 19:18:48.077151 18173 sql/temporary_schema.go:595  [n3] temporary object cleaner next scheduled to run at 2020-06-16 19:48:48.074833464 +0000 UTC
I200616 19:18:48.084599 16901 server/server_sql.go:645  [n3] done ensuring all necessary migrations have run
I200616 19:18:48.084679 16901 server/server.go:1756  [n3] serving sql connections
I200616 19:18:48.085827 18484 sqlmigrations/migrations.go:727  [n3] starting wait for upgrade finalization before schema change job migration
I200616 19:18:48.089445 18484 sqlmigrations/migrations.go:753  [n3] schema change job migration already complete
I200616 19:18:48.122723 18171 server/server_update.go:55  [n3] no need to upgrade, cluster already at the newest version
I200616 19:18:48.123099 18172 sql/event_log.go:132  [n3] Event: "node_join", target: 3, info: {Descriptor:{NodeID:3 Address:127.0.0.1:45571 Attrs: Locality:region=test,dc=dc3 ServerVersion:20.1-7 BuildTag:v20.2.0-alpha.00000000-2370-g82538176c8 StartedAt:1592335128031761398 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:46365} ClusterID:60994769-e235-4245-86bb-73c57939ea52 StartedAt:1592335128031761398 LastUp:1592335128031761398}
I200616 19:18:48.138098 18481 sql/event_log.go:132  [n1,client=127.0.0.1:50040,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I200616 19:18:48.199836 18481 sql/event_log.go:132  [n1,client=127.0.0.1:50040,hostssl,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I200616 19:18:48.210005 18481 sql/event_log.go:132  [n1,client=127.0.0.1:50040,hostssl,user=root] Event: "create_table", target: 53, info: {TableName:test.public.t Statement:CREATE TABLE test.public.t (x INT8 PRIMARY KEY) User:root}
I200616 19:18:48.231128 18481 kv/kvserver/replica_command.go:413  [n1,s1,r34/1:/{Table/38-Max}] initiating a split of this range at key /Table/53/1/6 [r35] (manual)
I200616 19:18:48.244941 18481 kv/kvserver/replica_command.go:1605  [n1,s1,r35/1:/{Table/53/1/6-Max}] change replicas (add [(n2,s2):2LEARNER] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, next=2, gen=1, sticky=9223372036.854775807,2147483647]
I200616 19:18:48.248018 18481 kv/kvserver/replica_raft.go:247  [n1,s1,r35/1:/{Table/53/1/6-Max}] proposing SIMPLE(l2) ADD_REPLICA[(n2,s2):2LEARNER]: after=[(n1,s1):1 (n2,s2):2LEARNER] next=3
I200616 19:18:48.250381 18481 kv/kvserver/store_snapshot.go:988  [n1,s1,r35/1:/{Table/53/1/6-Max}] streamed LEARNER snapshot f8ba66d2 at applied index 14 to (n2,s2):2LEARNER in 0.00s @ 0 B/s: kv pairs: 11, log entries: 0, rate-limit: 8.0 MiB/s, queued: 0.00s
I200616 19:18:48.250416 18852 kv/kvserver/raft_snapshot_queue.go:128  [n1,raftsnapshot,s1,r35/1:/{Table/53/1/6-Max}] skipping snapshot; replica is likely a learner in the process of being added: (n2,s2):2LEARNER
I200616 19:18:48.251057 18818 kv/kvserver/replica_raftstorage.go:791  [n2,s2,r35/2:{-}] applying LEARNER snapshot [id=f8ba66d2 index=14]
I200616 19:18:48.253022 18818 kv/kvserver/replica_raftstorage.go:812  [n2,s2,r35/2:/{Table/53/1/6-Max}] applied LEARNER snapshot [total=2ms ingestion=4@1ms id=f8ba66d2 index=14]
I200616 19:18:48.253834 18481 kv/kvserver/replica_command.go:1605  [n1,s1,r35/1:/{Table/53/1/6-Max}] change replicas (add [(n2,s2):2] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, (n2,s2):2LEARNER, next=3, gen=2, sticky=9223372036.854775807,2147483647]
I200616 19:18:48.256899 18481 kv/kvserver/replica_raft.go:247  [n1,s1,r35/1:/{Table/53/1/6-Max}] proposing SIMPLE(v2) ADD_REPLICA[(n2,s2):2]: after=[(n1,s1):1 (n2,s2):2] next=3
I200616 19:18:48.310053 18619 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove [(n1,s1):1VOTER_DEMOTING]): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1, (n2,s2):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]
I200616 19:18:48.317336 18619 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing ENTER_JOINT(r1 l1) REMOVE_REPLICA[(n1,s1):1VOTER_DEMOTING]: after=[(n1,s1):1VOTER_DEMOTING (n2,s2):2] next=3
I200616 19:18:48.319849 18619 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove []): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1VOTER_DEMOTING, (n2,s2):2, next=3, gen=4, sticky=9223372036.854775807,2147483647]
I200616 19:18:48.325813 18619 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing LEAVE_JOINT: after=[(n1,s1):1LEARNER (n2,s2):2] next=3
I200616 19:18:48.326965 18619 kv/kvserver/replica_command.go:1605  [n2,s2,r35/2:/{Table/53/1/6-Max}] change replicas (add [] remove [(n1,s1):1LEARNER]): existing descriptor r35:/{Table/53/1/6-Max} [(n1,s1):1LEARNER, (n2,s2):2, next=3, gen=5, sticky=9223372036.854775807,2147483647]
I200616 19:18:48.335513 18619 kv/kvserver/replica_raft.go:247  [n2,s2,r35/2:/{Table/53/1/6-Max}] proposing SIMPLE(r1) REMOVE_REPLICA[(n1,s1):1LEARNER]: after=[(n2,s2):2] next=3
I200616 19:18:48.336363 16975 kv/kvserver/store_remove_replica.go:122  [n1,s1,r35/1:/{Table/53/1/6-Max}] removing replica r35/1
I200616 19:18:48.342739 18481 sql/event_log.go:132  [n1,client=127.0.0.1:50040,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.defaults.results_buffer.size Value:0 User:root}
--- FAIL: TestDrainingProcessorSwallowsUncertaintyError (1.03s)
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=false
        --- FAIL: TestDrainingProcessorSwallowsUncertaintyError/vectorize=true/dummy=false (0.01s)
            processors_test.go:671: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1592335128.514995279,0 encountered previous write with future timestamp 1592335128.514995280,0 within uncertainty interval `t <= 1592335129.014995279,0`; observed timestamps: [{1 1592335128.514995279,0}] (SQLSTATE 40001)
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=true
    --- FAIL: TestDrainingProcessorSwallowsUncertaintyError/vectorize=true (0.01s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDrainingProcessorSwallowsUncertaintyError PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDrainingProcessorSwallowsUncertaintyError.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/materializer.go
+++ b/pkg/sql/colexec/materializer.go
@@ -34,6 +34,8 @@ type Materializer struct {
 
 	da sqlbase.DatumAlloc
 
+	drainHelper *drainHelper
+
 	// runtime fields --
 
 	// curIdx represents the current index into the column batch: the next row the
@@ -66,6 +68,61 @@ type Materializer struct {
 	closers []IdempotentCloser
 }
 
+// drainHelper is a utility struct that wraps MetadataSources in a RowSource
+// interface. This is done so that the Materializer can drain MetadataSources
+// in the vectorized input tree as inputs, rather than draining them in the
+// trailing metadata state, which is meant only for internal metadata
+// generation.
+type drainHelper struct {
+	execinfrapb.MetadataSources
+	ctx          context.Context
+	bufferedMeta []execinfrapb.ProducerMetadata
+}
+
+var _ execinfra.RowSource = &drainHelper{}
+
+func newDrainHelper(sources execinfrapb.MetadataSources) *drainHelper {
+	return &drainHelper{
+		MetadataSources: sources,
+	}
+}
+
+// OutputTypes implements the RowSource interface.
+func (d *drainHelper) OutputTypes() []*types.T {
+	colexecerror.InternalError("unimplemented")
+	// Unreachable code.
+	return nil
+}
+
+// Start implements the RowSource interface.
+func (d *drainHelper) Start(ctx context.Context) context.Context {
+	d.ctx = ctx
+	return ctx
+}
+
+// Next implements the RowSource interface.
+func (d *drainHelper) Next() (sqlbase.EncDatumRow, *execinfrapb.ProducerMetadata) {
+	if d.bufferedMeta == nil {
+		d.bufferedMeta = d.DrainMeta(d.ctx)
+		if d.bufferedMeta == nil {
+			// Still nil, avoid more calls to DrainMeta.
+			d.bufferedMeta = []execinfrapb.ProducerMetadata{}
+		}
+	}
+	if len(d.bufferedMeta) == 0 {
+		return nil, nil
+	}
+	meta := d.bufferedMeta[0]
+	d.bufferedMeta = d.bufferedMeta[1:]
+	return nil, &meta
+}
+
+// ConsumerDone implements the RowSource interface.
+func (d *drainHelper) ConsumerDone() {}
+
+// ConsumerClosed implements the RowSource interface.
+func (d *drainHelper) ConsumerClosed() {}
+
 const materializerProcName = "materializer"
 
 // NewMaterializer creates a new Materializer processor which processes the
@@ -93,10 +150,11 @@ func NewMaterializer(
 	cancelFlow func() context.CancelFunc,
 ) (*Materializer, error) {
 	m := &Materializer{
-		input:   input,
-		typs:    typs,
-		row:     make(sqlbase.EncDatumRow, len(typs)),
-		closers: toClose,
+		input:       input,
+		typs:        typs,
+		drainHelper: newDrainHelper(metadataSourcesQueue),
+		row:         make(sqlbase.EncDatumRow, len(typs)),
+		closers:     toClose,
 	}
 
 	if err := m.ProcessorBase.Init(
@@ -110,13 +168,10 @@ func NewMaterializer(
 		output,
 		nil, /* memMonitor */
 		execinfra.ProcStateOpts{
+			InputsToDrain: []execinfra.RowSource{m.drainHelper},
 			TrailingMetaCallback: func(ctx context.Context) []execinfrapb.ProducerMetadata {
-				var trailingMeta []execinfrapb.ProducerMetadata
-				for _, src := range metadataSourcesQueue {
-					trailingMeta = append(trailingMeta, src.DrainMeta(ctx)...)
-				}
 				m.InternalClose()
-				return trailingMeta
+				return nil
 			},
 		},
 	); err != nil {
@@ -147,6 +202,7 @@ func (m *Materializer) Child(nth int, verbose bool) execinfra.OpNode {
 // Start is part of the execinfra.RowSource interface.
 func (m *Materializer) Start(ctx context.Context) context.Context {
 	m.input.Init()
+	ctx = m.drainHelper.Start(ctx)
 	return m.ProcessorBase.StartInternal(ctx, materializerProcName)
 }
 

--- a/pkg/sql/colexec/parallel_unordered_synchronizer.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer.go
@@ -12,6 +12,7 @@ package colexec
 
 import (
 	"context"
+	"strings"
 	"sync"
 	"sync/atomic"
 
@@ -183,7 +184,18 @@ func (s *ParallelUnorderedSynchronizer) setState(state parallelUnorderedSynchron
 // affected by slow inputs.
 func (s *ParallelUnorderedSynchronizer) init(ctx context.Context) {
 	s.setState(parallelUnorderedSynchronizerStateRunning)
-	ctx, s.cancelFn = contextutil.WithCancel(ctx)
+	var (
+		cancelFn context.CancelFunc
+		// internalCancellation is an atomic that will be set to 1 if cancelFn is
+		// called (i.e. this is an internal cancellation), so that input goroutines
+		// know not propagate this cancellation.
+		internalCancellation int32
+	)
+	ctx, cancelFn = contextutil.WithCancel(ctx)
+	s.cancelFn = func() {
+		atomic.StoreInt32(&internalCancellation, 1)
+		cancelFn()
+	}
 	for i, input := range s.inputs {
 		s.nextBatch[i] = func(input SynchronizerInput, inputIdx int) func() {
 			return func() {
@@ -204,6 +216,10 @@ func (s *ParallelUnorderedSynchronizer) init(ctx context.Context) {
 				s.externalWaitGroup.Done()
 			}()
 			sendErr := func(err error) {
+				if strings.Contains(err.Error(), context.Canceled.Error()) && atomic.LoadInt32(&internalCancellation) == 1 {
+					// Don't propagate an internal cancellation error.
+					return
+				}
 				select {
 				// Non-blocking write to errCh, if an error is present the main
 				// goroutine will use that and cancel all inputs.

--- a/pkg/sql/rowexec/processors_test.go
+++ b/pkg/sql/rowexec/processors_test.go
@@ -624,8 +624,6 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 			}
 			vectorizeMode := "off"
 			if vectorize {
-				// TODO(yuzefovich): once #50299 is resolved, unskip this.
-				t.Skip("#50299")
 				vectorizeMode = "on"
 			}
 

